### PR TITLE
[core] SequenceController in SamplingParams

### PR DESCRIPTION
**Pull request by mmoskal**
This is an early work in progress.

This introduces a `controller: SequenceController` field in `SamplingParams`. This can be thought of as an extension of the already existing `logits_processors` (and in fact will work together with it).

The idea is to allow computation of logit token masks in the background, while the GPU is busy with the forward step, as well as [token healing](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38), [accelerated decoding](https://github.com/guidance-ai/guidance?tab=readme-ov-file#guidance-acceleration) (aka fast-forward or zero entropy tokens), and hidden tokens (typically stop tokens). This is essentially what's needed for running Guidance stateless grammars on the server.

@simon-mo please let me know if this looks like it's on the right track!

## Fast-forward tokens

The fast-forward tokens are handled naturally with BlockManagerV2 since they are used in prefix caching and chunked prefill. They can greatly speed generation of certain data formats (eg., JSON constrained by a schema).

## Backtracking

Large parts of this PR deal with backtracking, that is removing a bunch of recently generated tokens.

It is used by token healing in Guidance - the constraints sometimes make the model output tokens in un-natural way. Guidance periodically re-tokenizes the output (bytes) so far with canonical tokenizer. In some cases this requires removing a bunch of tokens from the output and re-inserting new (byte-equivalent) tokens.

Another motivation for this is stop tokens in Guidance - for example `f'Here's a joke: {gen(stop="\n")}; Score: {gen(regex=r"\d")}'` - here the LLM generates a joke `"Something funny\n"` and stops at a newline, however when generating score it should see `"Here's a joke: Something funny; Score: "`.

~Certain implementations of token healing also require it (though not the current one in Guidance).~ (edited text above - Guidance actually uses this)

Another motivation is when you want the LLM to do several independent, related tasks - after the context prompt, you ask for the first task, generate answer, backtrack away the answer and first task question, and continue. This is sometimes also used for XPIA mitigations, where you want to hide untrusted input when generating certain tokens.

Current implementation also uses backtrack when the controller requests a "no-op" (i.e., a backtrack of 0 with empty tokens to append) - this typically means that the controller run out of time computing token masks and wants to sit out a step (this shouldn't happen very often, but would otherwise block other sequences).

The reason backtracking is not a separate PR is that the SequenceController will allow for unit tests for it.

## AICI and Guidance

[AICI](https://github.com/microsoft/aici) (minus forks) can be implemented on top of this, so this supersedes #2888. There's an ongoing work on Guidance Controller in AICI.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by tdene** [comment]
[sampling_params imports sequence_controller](https://github.com/mmoskal/vllm/blob/seq_controller/vllm/sampling_params.py#L11)

[sequence_controller imports sequence](https://github.com/mmoskal/vllm/blob/seq_controller/vllm/sequence_controller.py#L3)

[sequence imports both sampling_params and sequence_controller](https://github.com/mmoskal/vllm/blob/seq_controller/vllm/sequence.py#L11)

Leading to Python circular import errors. Applying [PEP484](https://peps.python.org/pep-0484/#forward-references) to `sequence_controller.py` is a possible way to patch this.

**Comment by mmoskal** [comment]
closing in favor of #6273
